jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/virtuallist/JmixVirtualList.java

```diff
@@ -0,0 +1,122 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.jmix.flowui.component.virtuallist;
+
+import com.vaadin.flow.component.virtuallist.VirtualList;
+import com.vaadin.flow.data.provider.DataProvider;
+import com.vaadin.flow.data.provider.InMemoryDataProvider;
+import io.jmix.core.MetadataTools;
+import io.jmix.core.common.util.Preconditions;
+import io.jmix.core.metamodel.datatype.EnumClass;
+import io.jmix.flowui.component.delegate.VirtualListDelegate;
+import io.jmix.flowui.data.SupportsDataProvider;
+import io.jmix.flowui.data.SupportsItemsContainer;
+import io.jmix.flowui.data.SupportsItemsEnum;
+import io.jmix.flowui.data.items.ContainerDataProvider;
+import io.jmix.flowui.data.items.EnumDataProvider;
+import io.jmix.flowui.data.items.InMemoryDataProviderWrapper;
+import io.jmix.flowui.model.CollectionContainer;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.lang.Nullable;
+
+public class JmixVirtualList<V> extends VirtualList<V> implements SupportsDataProvider<V>,
+        SupportsItemsContainer<V>, SupportsItemsEnum<V>, ApplicationContextAware, InitializingBean {
+
+    protected ApplicationContext applicationContext;
+    protected MetadataTools metadataTools;
+
+    protected VirtualListDelegate<V> delegate;
+
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+        this.applicationContext = applicationContext;
+    }
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        autowireDependencies();
+        initComponent();
+    }
+
+    protected void autowireDependencies() {
+        metadataTools = applicationContext.getBean(MetadataTools.class);
+    }
+
+    protected void initComponent() {
+        delegate = createDelegate();
+        setRenderer(v -> metadataTools.format(v));
+    }
+
+    @SuppressWarnings("unchecked")
+    protected VirtualListDelegate<V> createDelegate() {
+        return applicationContext.getBean(VirtualListDelegate.class, this);
+    }
+
+    @Override
+    public void setItems(CollectionContainer<V> container) {
+        Preconditions.checkNotNullArgument(container);
+
+        setDataProvider(new ContainerDataProvider<>(container));
+    }
+
+    @Override
+    public void setItems(Class<V> itemsEnum) {
+        Preconditions.checkNotNullArgument(itemsEnum);
+
+        if (!itemsEnum.isEnum() || !EnumClass.class.isAssignableFrom(itemsEnum)) {
```

Comment:
In other component, this code is located in a delegate

Reply:
The delegate is used for JmixVirtualList only so there is no need to put it to delegate (other components use some common delegate)